mtgjson5/cache_builder.py

```
@@ -0,0 +1,429 @@
+        self.card_market_provider = CardMarketProvider()._get_card_mmarket_data()
+        self.multiverse_bridge_map = (
+            MultiverseBridgeProvider().get_rosetta_stone_cards()
+        )
+        self.cardhoarder_map = CardMarketProvider().get_cardmarket_price_map()
+
+        self.uuid_cache_df = self._load_uuid_cache()
+
+    def _load_uuid_cache(self) -> pl.DataFrame | None:
+        """Load legacy UUID mappings if available."""
+        cache_path = Path("resources/legacy_mtgjson_v5_uuid_mapping.json")
+        if not cache_path.exists():
+            return None
+        return pl.read_json(cache_path)
+
+    def get_set(self, set_code: str) -> dict[str, Any] | None:
+        """Get set metadata by set code."""
+        return self.sets_map.get(set_code.upper())
+
+    def get_foreign_data(
+        self, set_code: str, collector_number: str
+    ) -> list[dict[str, Any]]:
+        """Get foreign language data for a specific card."""
+        if self.foreign_data_map is None:
+            return []
+
+        result = self.foreign_data_map.filter(
+            (pl.col("set") == set_code.upper())
+            & (pl.col("collector_number") == collector_number)
+        )
+
+        if result.is_empty():
+            return []
+
+        # Extract and force to Python list
+        foreign_data = result["foreign_data"].to_list()[0]
+
+        if foreign_data is None:
+            return []
+
+        return list(foreign_data)
+
+    def get_meld_parts(self, result_name: str) -> list[str] | None:
+        """Get meld part names for a meld result card."""
+        for triplet in self.meld_triplets:
+            if triplet[2] == result_name:
+                return triplet[:2]
+        return None
+
+    @classmethod
+    def get_instance(cls) -> "GlobalCache":
+        """Get the singleton instance of GlobalCache."""
+        if cls._instance is None:
+            cls()
+        assert cls._instance is not None
```

asserts should not be in production code, only test code. You can throw an exception if there's a problem, but don't use assert